### PR TITLE
ref(search): dont parse token values

### PR DESCRIPTION
**Pull request by JonasBa**
remove rawValue in favor of parsed and make sure we parse values correctly for tokens like relative date. I also noticed that the grammar for size units only recognizes lowercase inputs, so I'm making those case insensitive.

I've made parsing a config option as most parts of the app will not require it.

### CODE REVIEW

File: static/app/components/searchSyntax/parser.tsx [hunk 1]
```diff
@@ -968,10 +1056,15 @@ export type SearchConfig = {
...
   getFilterTokenWarning?: (key: string) => React.ReactNode;
+  /**
+   * Determines if user input values should be parsed
+   */
+  parse?: boolean;
   /**
    * If validateKeys is set to true, tag keys that don't exist in supportedTags will be consider invalid
    */
   supportedTags?: TagCollection;
+
```
Comment (evanpurkhiser) [inline static/app/components/searchSyntax/parser.tsx:1067]: nit: extranious blank line

File: static/app/components/searchSyntax/parser.tsx [hunk 2]
```diff
@@ -878,6 +890,82 @@ export class TokenConverter {
...
+    default:
+      throw new Error('Invalid unit');
+  }
+
+  if (offset === undefined) {
+    throw new Error('Unreachable');
+  }
+
+  date = sign === '-' ? date - offset : date + offset;
+  return new Date(date);
```
Comment (evanpurkhiser) [inline static/app/components/searchSyntax/parser.tsx:936]: Should this actually produce a moment `Difference` object?

File: static/app/components/searchSyntax/parser.tsx [hunk 3]
```diff
@@ -878,6 +890,82 @@ export class TokenConverter {
...
+function numeric(input: string) {
+  const number = parseFloat(input);
+  if (isNaN(number)) {
+    throw new Error('Invalid number');
+  }
+  return number;
+}
+function parseDuration(input: string, _unit: string): number {
+  return numeric(input);
+}
```
Comment (evanpurkhiser) [inline static/app/components/searchSyntax/parser.tsx:949]: Why isn't the unit used?

File: static/app/components/searchSyntax/grammar.pegjs [hunk 4]
```diff
@@ -345,18 +346,24 @@ rel_date_format
...
   = value:numeric
-    unit:("bit"/"nb"/"bytes"/"kb"/"mb"/"gb"/"tb"/"pb"/"eb"/"zb"/"yb"/"kib"/"mib"/"gib"/"tib"/"pib"/"eib"/"zib"/"yib")
+    unit:(size_unit)
     &end_value {
       return tc.tokenValueSize(value, unit);
     }
 
+size_unit = bits / bytes
+bits = "bit"i / "kib"i / "mib"i / "gib"i / "tib"i / "pib"i / "eib"i / "zib"i / "yib"i
+bytes = "bytes"i / "nb"i / "kb"i / "mb"i / "gb"i / "tb"i / "pb"i / "eb"i / "zb"i / "yb"i
```
Comment (evanpurkhiser) [inline static/app/components/searchSyntax/grammar.pegjs:365]: 👍 for these changes, would actually not mind seeing these pulled out into a separate PR

File: static/app/components/searchSyntax/parser.tsx (hunk 2 above)
Comment (JonasBa) [reply static/app/components/searchSyntax/parser.tsx:936]: I dont think so, I'd rather just have it produce a point in time, the usage is likely going to be a comparison of getTime() between two dates. I wouldnt want to force the usage of moment here

File: fixtures/search-syntax/timestamp.json [hunk 5]
```diff
@@ -10,7 +10,7 @@
         "negated": false,
         "key": {"type": "keySimple", "value": "timestamp", "quoted": false},
         "operator": ">",
-        "value": {"type": "valueIso8601Date", "value": "2015-05-18T00:00:00.000Z"}
+        "value": {"type": "valueIso8601Date", "value": "2015-05-18"}
```
Comment (evanpurkhiser) [inline fixtures/search-syntax/timestamp.json:13]: This is still an ISO8601 date right?

File: fixtures/search-syntax/supported_tags.json [hunk 6]
```diff
@@ -72,7 +72,7 @@
         "operator": ">",
         "value": {
           "type": "valueNumber",
-          "rawValue": 0,
+
```
Comment (evanpurkhiser) [inline fixtures/search-syntax/supported_tags.json:75]: nit: blank line

File: static/app/components/searchSyntax/parser.tsx (hunk 2 above)
Comment (evanpurkhiser) [reply static/app/components/searchSyntax/parser.tsx:936]: We could also maybe just represent this as + - milliseconds. Then users if they want it can do

```tsx
new Date(new Date().getTime() + difference)
```

File: fixtures/search-syntax/specific_time_filter.json [hunk 7]
```diff
@@ -74,7 +74,7 @@
         "operator": "",
         "value": {
           "type": "valueIso8601Date",
-          "value": "2018-01-01T05:06:07.000Z"
+          "value": "2018-01-01T05:06:07+00:00"
```
[the remark this answers is not in the document]
Comment (JonasBa) [reply fixtures/search-syntax/specific_time_filter.json:77]: I think both representations here are correct, JS just seems to coerce to that format

File: static/app/components/searchSyntax/parser.tsx (hunk 3 above)
Comment (JonasBa) [reply static/app/components/searchSyntax/parser.tsx:949]: We would need to make a decision of what unit to what base unit we want to normalize to. This just parses the value so that whoever is consuming this can make the decision of what base to normalize this to

File: static/app/components/searchSyntax/grammar.pegjs (hunk 4 above)
Comment (JonasBa) [reply static/app/components/searchSyntax/grammar.pegjs:365]: Yup, lets do that

File: static/app/components/searchSyntax/parser.tsx (hunk 3 above)
Comment (evanpurkhiser) [reply static/app/components/searchSyntax/parser.tsx:949]: we should just parse down to milliseconds no?

File: src/sentry/api/event_search.py [hunk 8]
```diff
@@ -169,9 +169,14 @@
 iso_8601_date_format = date_format time_format? ("Z" / tz_format)? &end_value
 rel_date_format      = ~r"[+-][0-9]+[wdhm]" &end_value
 duration_format      = numeric ("ms"/"s"/"min"/"m"/"hr"/"h"/"day"/"d"/"wk"/"w") &end_value
```
Comment (evanpurkhiser) [inline src/sentry/api/event_search.py:171]: Can we match what we did in the pegjs grammar and pull out the `duration_unit` too? I was trying to keep these as close to the same as possible

File: static/app/components/searchSyntax/parser.tsx (hunk 2 above)
Comment (JonasBa) [reply static/app/components/searchSyntax/parser.tsx:936]: Since the token has a date type, I'd rather represent it as a date tbh

File: static/app/components/searchSyntax/parser.tsx (hunk 3 above)
Comment (JonasBa) [reply static/app/components/searchSyntax/parser.tsx:949]: Yes, but I cant say if thats what everyone would want, so I'd rather not make that decision at all and keep it open. We would need to make the equivalent decisions for sizes and other units and I'm not sure yet if it's something we can make a decision about

Comment (evanpurkhiser) [reply static/app/components/searchSyntax/parser.tsx:949]: This feels more right to me to normalize it though. I don't want to have to do the work to parse the units, as a user of the token results I would expect the parsed value to be normalized down.

IMO we should be doing the same for size units, it should be parsed down to bytes

Comment (JonasBa) [reply static/app/components/searchSyntax/parser.tsx]: Yeah, I agree, it's just that I dont know if I know enough to make that call 😂

File: src/sentry/api/event_search.py (hunk 8 above)
Comment (JonasBa) [reply src/sentry/api/event_search.py]: Yup, I'll make all these units separate expression in the next PR when I update the grammar to not be case sensitive

**Review by evanpurkhiser — APPROVED** [review]
